File: scripts/seed-zip-centroids.py
#!/usr/bin/env python3
"""
Loads US zip-code centroids from supabase/seed/us_zip_centroids.csv into the
`zip_centroids` table (created by supabase/migrations/019_zip_centroids.sql).

Usage (from repo root):
  python3 scripts/seed-zip-centroids.py

Reads SUPABASE_URL and SUPABASE_SERVICE_KEY from (in order of priority):
  process environment → apps/gatsby-glass/.env.local → .env.local → .env
"""

import csv
import json
import os
import sys
import urllib.request
import urllib.error
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent
CSV_PATH  = REPO_ROOT / "supabase" / "seed" / "us_zip_centroids.csv"
BATCH_SIZE = 1000


def load_env_files():
    candidates = [
        REPO_ROOT / ".env.local",
        REPO_ROOT / ".env",
        REPO_ROOT / "apps" / "gatsby-glass" / ".env.local",
        REPO_ROOT / "apps" / "gatsby-glass" / ".env",
    ]
    for path in candidates:
        if not path.exists():
            continue
        with open(path, encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith("#"):
                    continue
                if "=" not in line:
                    continue
                key, _, val = line.partition("=")
                key = key.strip()
                val = val.strip().strip("\"'")
                if key and key not in os.environ:
                    os.environ[key] = val


def parse_csv(path):
    rows = []
    with open(path, newline="", encoding="utf-8") as f:
        for line in f:
            stripped = line.strip()
            if not stripped or stripped.startswith("#"):
                continue
            break  # first non-comment line is the header

        reader = csv.DictReader(f, fieldnames=["zip_code","lat","lng","city","state"])
        for row in reader:
            z = (row.get("zip_code") or "").strip()
            lat_s = (row.get("lat") or "").strip()
            lng_s = (row.get("lng") or "").strip()
            if not z or z == "zip_code":
                continue
            if len(z) != 5 or not z.isdigit():
                continue
            try:
                lat = float(lat_s)
                lng = float(lng_s)
            except ValueError:
                continue
            rows.append({
                "zip_code": z,
                "lat": lat,
                "lng": lng,
                "city": (row.get("city") or "").strip() or None,
                "state": (row.get("state") or "").strip() or None,
            })
    return rows


def upsert_batch(url, key, batch):
    endpoint = f"{url.rstrip('/')}/rest/v1/zip_centroids"
    data = json.dumps(batch).encode("utf-8")
    req = urllib.request.Request(
        endpoint,
        data=data,
        method="POST",
        headers={
            "apikey": key,
            "Authorization": f"Bearer {key}",
            "Content-Type": "application/json",
            "Prefer": "resolution=merge-duplicates,return=minimal",
        },
    )
    try:
        with urllib.request.urlopen(req) as resp:
            if resp.status not in (200, 201, 204):
                body = resp.read().decode()
                print(f"Unexpected status {resp.status}: {body}", file=sys.stderr)
                sys.exit(1)
    except urllib.error.HTTPError as e:
        body = e.read().decode()
        print(f"HTTP {e.code}: {body}", file=sys.stderr)
        sys.exit(1)


def main():
    load_env_files()

    supabase_url = os.environ.get("SUPABASE_URL") or os.environ.get("NEXT_PUBLIC_SUPABASE_URL", "")
    supabase_key = os.environ.get("SUPABASE_SERVICE_KEY", "")

    if not supabase_url or not supabase_key:
        print("Missing SUPABASE_URL and/or SUPABASE_SERVICE_KEY.", file=sys.stderr)
        print("Set them in apps/gatsby-glass/.env.local", file=sys.stderr)
        sys.exit(1)

    if "PASTE_YOUR_SERVICE" in supabase_key:
        print("SUPABASE_SERVICE_KEY is still the placeholder.", file=sys.stderr)
        print("Paste your service_role / secret key in apps/gatsby-glass/.env.local", file=sys.stderr)
        sys.exit(1)

    rows = parse_csv(CSV_PATH)
    if not rows:
        print(f"No valid rows parsed from {CSV_PATH}", file=sys.stderr)
        sys.exit(1)

    print(f"Parsed {len(rows)} zip centroids from {CSV_PATH}")

    upserted = 0
    for i in range(0, len(rows), BATCH_SIZE):
        batch = rows[i:i + BATCH_SIZE]
        upsert_batch(supabase_url, supabase_key, batch)
        upserted += len(batch)
        print(f"Upserted {upserted}/{len(rows)}")

    print(f"Done. {upserted} zip centroids loaded.")


if __name__ == "__main__":
    main()

File: scripts/seed-zip-centroids.mjs
#!/usr/bin/env node
// Loads US zip-code centroids from supabase/seed/us_zip_centroids.csv into the
// `zip_centroids` table (created by supabase/migrations/019_zip_centroids.sql).
//
// The centroids power the territory mileage buffer precomputed by
// sync_gatsby_glass_locations() and the nearest_franchise() RPC used to route
// out-of-territory quote leads.
//
// Usage (from repo root), after running migration 019:
//   SUPABASE_URL=... SUPABASE_SERVICE_KEY=... node scripts/seed-zip-centroids.mjs
//
// The CSV header must be: zip_code,lat,lng,city,state
// Lines starting with '#' and blank lines are ignored. Re-running is safe —
// rows are upserted on the zip_code primary key.
//
// Implemented against the Supabase REST (PostgREST) endpoint with the service
// key so it has no npm dependencies and runs with a bare `node` (>=18).

import { existsSync, readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = resolve(__dirname, '..');
const CSV_PATH = resolve(REPO_ROOT, 'supabase/seed/us_zip_centroids.csv');
const BATCH_SIZE = 1000;

// Lightweight .env loader (no dependency). Loads the first matching file so a
// plain `node`/`pnpm seed:zip-centroids` works without exporting vars by hand.
// Existing process.env values always win.
function loadEnvFiles() {
  const candidates = [
    resolve(REPO_ROOT, '.env.local'),
    resolve(REPO_ROOT, '.env'),
    resolve(REPO_ROOT, 'apps/gatsby-glass/.env.local'),
    resolve(REPO_ROOT, 'apps/gatsby-glass/.env'),
  ];
  for (const file of candidates) {
    if (!existsSync(file)) continue;
    const text = readFileSync(file, 'utf8');
    for (const raw of text.split(/\r?\n/)) {
      const line = raw.trim();
      if (!line || line.startsWith('#')) continue;
      const eq = line.indexOf('=');
      if (eq === -1) continue;
      const key = line.slice(0, eq).trim();
      let value = line.slice(eq + 1).trim();
      // Strip surrounding quotes if present.
      if (
        (value.startsWith('"') && value.endsWith('"')) ||
        (value.startsWith("'") && value.endsWith("'"))
      ) {
        value = value.slice(1, -1);
      }
      if (key && process.env[key] === undefined) {
        process.env[key] = value;
      }
    }
  }
}

loadEnvFiles();

const supabaseUrl = (
  process.env.SUPABASE_URL ||
  process.env.NEXT_PUBLIC_SUPABASE_URL ||
  ''
).replace(/\/$/, '');
const supabaseKey = process.env.SUPABASE_SERVICE_KEY;

if (!supabaseUrl || !supabaseKey) {
  console.error(
    'Missing SUPABASE_URL and/or SUPABASE_SERVICE_KEY environment variables.'
  );
  process.exit(1);
}

if (supabaseKey.includes('PASTE_YOUR_SERVICE')) {
  console.error(
    'SUPABASE_SERVICE_KEY is still the placeholder. Set the service_role / secret key\n' +
    '(Supabase Dashboard → Project Settings → API) in apps/gatsby-glass/.env.local, then re-run.'
  );
  process.exit(1);
}

function parseCsv(text) {
  const rows = [];
  const lines = text.split(/\r?\n/);
  for (const raw of lines) {
    const line = raw.trim();
    if (!line || line.startsWith('#')) continue;
    const cols = line.split(',');
    const [zip, lat, lng] = cols;
    // Skip the header row.
    if (zip === 'zip_code') continue;

    const zipClean = (zip || '').trim();
    const latNum = Number.parseFloat(lat);
    const lngNum = Number.parseFloat(lng);
    if (!/^\d{5}$/.test(zipClean) || Number.isNaN(latNum) || Number.isNaN(lngNum)) {
      continue;
    }

    rows.push({
      zip_code: zipClean,
      lat: latNum,
      lng: lngNum,
      city: (cols[3] || '').trim() || null,
      state: (cols[4] || '').trim() || null,
    });
  }
  return rows;
}

async function upsertBatch(batch) {
  const res = await fetch(`${supabaseUrl}/rest/v1/zip_centroids`, {
    method: 'POST',
    headers: {
      apikey: supabaseKey,
      Authorization: `Bearer ${supabaseKey}`,
      'Content-Type': 'application/json',
      // Upsert on the primary key (zip_code) and don't return rows.
      Prefer: 'resolution=merge-duplicates,return=minimal',
    },
    body: JSON.stringify(batch),
  });

  if (!res.ok) {
    const text = await res.text().catch(() => '');
    throw new Error(`HTTP ${res.status}: ${text}`);
  }
}

async function main() {
  const csv = readFileSync(CSV_PATH, 'utf8');
  const rows = parseCsv(csv);

  if (rows.length === 0) {
    console.error(`No valid rows parsed from ${CSV_PATH}`);
    process.exit(1);
  }

  console.log(`Parsed ${rows.length} zip centroids from ${CSV_PATH}`);

  let upserted = 0;
  for (let i = 0; i < rows.length; i += BATCH_SIZE) {
    const batch = rows.slice(i, i + BATCH_SIZE);
    await upsertBatch(batch);
    upserted += batch.length;
    console.log(`Upserted ${upserted}/${rows.length}`);
  }

  console.log(`Done. ${upserted} zip centroids loaded.`);
}

main().catch((err) => {
  console.error(err.message || err);
  process.exit(1);
});
